OrbitLinuxTracing/LinuxTracingUtils.cpp

```diff
@@ -35,40 +37,60 @@ std::string ReadMaps(pid_t pid) {
   std::optional<std::string> maps_content_opt = ReadFile(maps_filename);
   if (maps_content_opt.has_value()) {
     return maps_content_opt.value();
-  } else {
-    return "";
   }
+  return "";
 }
 
-std::optional<std::string> ExecuteCommand(const std::string& cmd) {
-  std::unique_ptr<FILE, decltype(&pclose)> pipe{popen(cmd.c_str(), "r"), pclose};
-  if (!pipe) {
-    ERROR("Could not open pipe for \"%s\"", cmd.c_str());
-    return std::optional<std::string>{};
+static std::optional<pid_t> ProcEntryToPid(const std::filesystem::directory_entry& entry) {
+  if (!entry.is_directory()) {
+    return std::nullopt;
   }
 
-  std::array<char, 128> buffer;
-  std::string result;
-  while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
-    result += buffer.data();
+  int potential_pid;
+  if (!absl::SimpleAtoi(entry.path().filename().string(), &potential_pid)) {
+    return std::nullopt;
   }
-  return result;
+
+  if (potential_pid <= 0) {
+    return std::nullopt;
+  }
+
+  return static_cast<pid_t>(potential_pid);
 }
 
-std::vector<pid_t> ListThreads(pid_t pid) {
-  std::vector<pid_t> threads;
-  std::optional<std::string> tasks = ExecuteCommand(absl::StrFormat("ls /proc/%d/task", pid));
-  if (!tasks.has_value()) {
-    return {};
+std::vector<pid_t> GetAllPids() {
+  fs::directory_iterator proc{"/proc"};
+  std::vector<pid_t> pids;
+
+  for (const auto& entry : proc) {
+    if (auto pid = ProcEntryToPid(entry); pid.has_value()) {
```

Comment:
Is there a reason to not move the declaration/assignment of pid out of the condition. This makes it harder to read.

Reply:
This is a nice C++17 feature. The point of it is to limit the scope of a variable used only in the if to only the if. In other reviews I've actually been told to use the feature, as I tend to forget about it even though I like it (this was actually Henning's code, that's why I "remembered" this time).

Reply:
Okay, personally I think it is harder to read, and the effective scope in those cases here, will be exactly the same, as the outer scope (of the loop) also ends with directly after the one of the if.
Anyways, feel free to decide whatever you like more.

And out of curiosity, is that really a C++17 feature? I had the feeling that its been there for years in Java, and the adaption rate there, is not very high. So I thought it would be also there since ever in C++. :)

Reply:
https://en.cppreference.com/w/cpp/language/if
Is this really in Java? I've never seen it.

Reply:
I think this is indeed a great feature that we should all get used to use more often, and it's also consistent with how declarations are done in for loops. 

> Okay, personally I think it is harder to read, and the effective scope in those cases here, will be exactly the same, as the outer scope (of the loop) also ends with directly after the one of the if.

Unless someone changes the code someday and adds something below the if. Better to reduce the scope to what is actually needed now.

Reply:
Thanks for the Link and for the explanation. Personally, I am still not convinced, though, but that doesn't matter for now.
And @dpallotti, you was correct. Java does not have this (at least Java 8), I wonder why I thought so. Pretty sure, I rarely saw this.